### PR TITLE
:sparkles: Refactor external objects & reduce code duplication

### LINKED ISSUE #2168: External watchers should key off of GroupKind, not GroupVersionKind
Related to #2140, if we have code like this:

```go
_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
```

we probably need to change it to key off of just GroupKind. Otherwise, if we see an object reference to 2 different API versions for the same kind (e.g. v1alpha2 and v1alpha3 KubeadmConfig), we'll start 2 different `source.Kind` watchers, 1 for each API version.

We'll likely need to tie in to the fix for #2140, because if we have a v1alpha3 controller and it sees a v1alpha2 external object, we'll want to make sure we watch the v1alpha3 objects and not v1alpha2.

cc @detiber @vincepri @chuckha 

/kind cleanup
/priority important-longterm
/milestone Next

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
/assign @ncdc 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2168

### CODE REVIEW

File: controllers/external/types.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2020 The Kubernetes Authors.
```
Comment (detiber) [inline controllers/external/types.go]: Revert back to 2019?

Comment (vincepri) [reply controllers/external/types.go]: This was actually created this year

File: controllers/external/tracker.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	Controller controller.Controller
+}
+
+// Watch uses the controller to issue a Watch only if the object hasn't been seen before.
+func (o *ObjectTracker) Watch(log logr.Logger, obj runtime.Object, handler handler.EventHandler) error {
+	gk := obj.GetObjectKind().GroupVersionKind().GroupKind()
+
+	_, loaded := o.m.LoadOrStore(gk.String(), struct{}{})
+	if loaded || o.Controller == nil {
```
Comment (fabriziopandini) [inline controllers/external/tracker.go]: I find `o.Controller == nil` test a little bit confusing here.
What about moving the test at the beginning of the func and returning error vs return nil

Comment (vincepri) [reply controllers/external/tracker.go]: Fixed, thanks!

**Comment by fabriziopandini** [comment]
/lgtm
Thanks @vincepri

**Comment by vincepri** [comment]
I think the capd job is optional, and @chuckha has been working on some fixes for it

**Comment by chuckha** [comment]
Yes, it's optional! It does provide nice signal, but you don't have to wait for it if you don't feel like it

**Comment by chuckha** [comment]
This is an interesting failure though. Not sure why deletes would be unhappy with this change set, but i've never seen this failure before.

**Comment by vincepri** [comment]
@chuckha That's a good point, not totally sure

**Pull request merged**